esp-hal/src/uart.rs

```diff
@@ -662,6 +664,42 @@ where
 
         Ok(uart_tx)
     }
+
+    /// Reconfigures the driver to operate in [`Async`] mode.
+    pub fn into_async(self) -> UartTx<'d, Async, T> {
+        if !self.uart.state().is_rx_async.load(Ordering::Acquire) {
+            self.uart
+                .info()
+                .set_interrupt_handler(self.uart.info().async_handler);
+        }
+        self.uart.state().is_tx_async.store(true, Ordering::Release);
+
+        UartTx {
+            uart: self.uart,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'d, T> UartTx<'d, Async, T>
+where
+    T: Instance,
+{
+    /// Reconfigures the driver to operate in [`Blocking`] mode.
+    pub fn into_blocking(self) -> UartTx<'d, Blocking, T> {
+        self.uart
```

Comment:
This should be fine as long as we never add a `set_interrupt_handler` impl for the split structs. `UartRx` could still be in async mode whilst we convert `UartTx` to blocking. We could then call `set_interrupt_handler` on `UartTx` and override the handler.

Though I do wonder if this is still worth it, I'm starting to come around to the idea of removing the split side constructors because when would you ever call `into_blocking` on a split half, given that any async mode will also impl the blocking methods/traits...

Reply:
> This should be fine as long as we never add a set_interrupt_handler impl for the split structs.

If we do, we can check the "other" async flag and panic or return an error.

Reply:
Removing the split constructors and separate into_asyncs would help, I wouldn't mind bypassing the "send RX to core 1 and turn it into Async, then turn TX into Async on core 0" problem - good luck to me hammering out edge cases :D

Reply:
True, let's leave this as is for now then.